File: Formal Languages and Compiler Design/L01/ProgramInternalForm.py
class ProgramInternalForm:
    def __init__(self):
        self.__list = []

    def add(self, tokenCode, stIndex=None):
        pair = (tokenCode, stIndex)
        self.__list.append(pair)

    def __str__(self):
        s = ''
        s += '| %-5s | %-5s |\n' % ('Code', 'Addr')
        s += '|-%-5s-+-%-5s-+\n' % ('-' * 5, '-' * 5)
        for entry in self.__list:
            s += '| %-5s | %-5s |\n' % (str(entry[0]), str(entry[1]))
        s += '|-%-5s-+-%-5s-+\n' % ('-' * 5, '-' * 5)
        return s

File: Formal Languages and Compiler Design/L02_2/src/scanner/ProgramInternalForm.java
package scanner;

import util.Pair;

import java.util.ArrayList;
import java.util.List;

public class ProgramInternalForm {
    private List<Pair<Integer, Integer>> list;

    public ProgramInternalForm() {
        list = new ArrayList<>();
    }

    public void add(Integer tokenCode, Integer indexST) {
        this.list.add(new Pair<>(tokenCode, indexST));
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append(String.format("| %-5s | %-5s |\n", "Code", "Addr"));
        sb.append(String.format("+-%-5s-+-%-5s-+\n", "-----", "-----"));
        for (Pair<Integer, Integer> el : this.list) {
            sb.append(String.format("| %-5s | %-5s |\n", el.getFirst(), el.getSecond()));
        }
        sb.append(String.format("+-%-5s-+-%-5s-+\n", "-----", "-----"));

        return sb.toString();
//        return "ProgramInternalForm{" +
//                "list=" + list +
//                '}';
    }
}
